contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -0,0 +1,406 @@
+information to other plugins. This library keeps a map from keys of type string
+to opaque pointers of type interface{}. A write operation takes a key and a
+pointer and stores the opaque pointer in the map with the given key. Other
+plugins can provide the key and receive the opaque pointer. Multiple plugins can
+share the state or communicate via this mechanism.  
+The saved state is preserved only during a single scheduling cycle. At the end
+of a scheduling cycle, this map is destructed. So, plugins cannot keep shared
+state across multiple scheduling cycle. They can, however, update the scheduler
+cache via the provided interface of the cache. The cache interface allows
+limited state preservation across multiple scheduling cycle.  
+It is worth noting that plugins are assumed to be **trusted**. Scheduler does
+not prevent one plugin from accessing or modifying another plugin's state.
+
+## Plugin registration
+
+Plugin registration is done by providing an extension point and a function that
+should be called at that extension point. This step will be something like:
+
+<table>
+<thead>
+<tr>
+<th><p><pre>
+register("pre-filter", plugin.foo)
+</pre></p>
+
+</th>
+</tr>
+</thead>
+<tbody>
+</tbody>
+</table>
+
+The details of the function signature will be provided later.
+
+## Extension points
+
+The following picture shows the scheduling cycle of a Pod and the extension
+points that the scheduling framework exposes. In this picture "Filter" is
+equivalent to "Predicate" in scheduler v1 and "Scoring" is equivalent to
+"Priority function". Plugins are go functions. They are registered to be called
+at one of these extension points. They are called by the framework in the same
+order they are registered for each extension point.  
+In the following sections we describe each extension point in the same order
+they are called in a schedule cycle.
+
+![image](images/scheduling-framework-extensions.png)
+
+### Scheduling queue sort
+
+These plugins indicate how Pods should be sorted in the scheduling queue. A
+plugin registered at this point only returns greater, smaller, or equal to
+indicate an ordering between two Pods. In other words, a plugin at this
+extension point returns the answer to "less(pod1, pod2)". Multiple plugins may
+be registered at this point. Plugins registered at this point are called in
+order and the invocation continues as long as plugins return "equal". Once a
+plugin returns "greater" or "smaller" the invocation of these plugins are
+stopped.
+
+### Pre-filter
+
+These plugins are generally useful to check certain conditions that the cluster
+or the Pod must meet. These are also useful to perform pre-processing on the pod
+and store some information about the pod that can be used by other plugins.  
+The pod pointer is passed as an argument to these plugins. If any of these
+plugins return an error, the scheduling cycle is aborted.  
+These plugins are called serially in the same order registered.
+
+### Filter
```

Comment:
Are there any plans on how to integrate this new framework with the predicate equivalence cache?

Reply:
What I think we should do is to provide an equivalence cache library in the framework and leave it up to the plugins to use it or not. In more details:
- The framework creates a structure similar to our equivalence cache.
- The framework provides an interface to interact with the eCache.
- Filter plugins can call a function to obtain a pointer to the cache and use its interface to store or retrieve its data. It will be up to the plugin to use the stored data or not.

Reply:
The main interaction between plugin and ecache I think would be the informer handlers to invalidate the predicates.

Reply:
I agree that it is also an important part of the interaction of plugins and eCache. Generally, the framework should allow plugins to register their own handler at various object events, such as pod update, node delete, etc. The framework calls their handles at the events.
Invalidating eCache is one of many operations that a plugin may want to perform at any of the events. Similarly, plugins may want to invalidate their other internal caches or update them at object events.